pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,572 @@
+     */
+    template <typename T>
+    void Scatter(T *sendBuf, T *recvBuf, size_t dataSize, int root) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        PL_MPI_IS_SUCCESS(MPI_Scatter(sendBuf, dataSize, datatype, recvBuf,
+                                      dataSize, datatype, root,
+                                      this->getComm()));
+    }
+
+    /**
+     * @brief MPI_Scatter wrapper.
+     *
+     * @tparam T C++ data type.
+     * @param sendBuf Send buffer vector.
+     * @param recvBuf Receive buffer vector.
+     * @param root Rank of scatter root.
+     */
+    template <typename T>
+    void Scatter(std::vector<T> &sendBuf, std::vector<T> &recvBuf, int root) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        PL_ABORT_IF(sendBuf.size() != recvBuf.size() * this->getSize(),
+                    "Incompatible size of sendBuf and recvBuf.");
+        PL_MPI_IS_SUCCESS(MPI_Scatter(sendBuf.data(), recvBuf.size(), datatype,
+                                      recvBuf.data(), recvBuf.size(), datatype,
+                                      root, this->getComm()));
+    }
+
+    /**
+     * @brief MPI_Scatter wrapper.
+     *
+     * @tparam T C++ data type.
+     * @param sendBuf Send buffer vector.
+     * @param root Rank of scatter root.
+     *
+     * @return recvBuf Receive buffer vector.
+     */
+    template <typename T>
+    auto scatter(std::vector<T> &sendBuf, int root) -> std::vector<T> {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        int recvBufSize = sendBuf.size() / this->getSize();
+        std::vector<T> recvBuf(recvBufSize);
+        PL_MPI_IS_SUCCESS(MPI_Scatter(sendBuf.data(), recvBuf.size(), datatype,
+                                      recvBuf.data(), recvBuf.size(), datatype,
+                                      root, this->getComm()));
+        return recvBuf;
+    }
+
+    /**
+     * @brief MPI_Scatter wrapper.
+     *
+     * @tparam T C++ data type.
+     * @param sendBuf Send buffer.
+     * @param dest Rank of destination.
+     * @param recvBuf Receive buffer.
+     * @param source Rank of source.
+     */
+    template <typename T>
+    void Sendrecv(T &sendBuf, int dest, T &recvBuf, int source) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Status status;
+        int sendtag = 0;
+        int recvtag = 0;
+        PL_MPI_IS_SUCCESS(MPI_Sendrecv(&sendBuf, 1, datatype, dest, sendtag,
+                                       &recvBuf, 1, datatype, source, recvtag,
+                                       this->getComm(), &status));
+    }
+
+    /**
+     * @brief MPI_Scatter wrapper.
+     *
+     * @tparam T C++ data type.
+     * @param sendBuf Send buffer vector.
+     * @param dest Rank of destination.
+     * @param recvBuf Receive buffer vector.
+     * @param source Rank of source.
+     */
+    template <typename T>
+    void Sendrecv(std::vector<T> &sendBuf, int dest, std::vector<T> &recvBuf,
```

Comment:
Here we could also `Sendrecv` the send buffer sizes and allocate the recv buffers to make this more general/robust.

Reply:
as discussed, we'll keep current implementation.